sqstaskmaster/queue_scale.py

```diff
@@ -84,26 +92,86 @@ def run(self):
 
                 # This could result in thrashing the service count. For tasks that run quickly some historical filter
-                # might be required
-                self.ecs.update_service(
-                    cluster=rule[self.CLUSTER_NAME],
-                    service=rule[self.SERVICE_NAME],
-                    desiredCount=rule[self.ACTIVE_SIZE] if total_messages > 0 else 0,
-                )
-
-                logger.info(
-                    "Setting: %s %s to %s count",
-                    rule[self.CLUSTER_NAME],
-                    rule[self.SERVICE_NAME],
-                    rule[self.ACTIVE_SIZE] if total_messages > 0 else 0,
-                )
+                # might be required but that is not the intended use case.
+                # Beware of attempting to set the count based on the number of messages in the queue. It will be
+                # difficult to control which instances get shutdown and ensure a graceful exit.
+                # All or nothing is simple and only wasteful at the tail end of a large process queue
+
+                current_state = self.service_state(rule)
+                if current_state == ServiceState.ACTIVE:
+                    self.ecs.update_service(
+                        cluster=rule[self.CLUSTER_NAME],
+                        service=rule[self.SERVICE_NAME],
+                        desiredCount=rule[self.ACTIVE_SIZE]
+                        if total_messages > 0
+                        else 0,
+                    )
+
+                    logger.info(
+                        "Setting: %s %s to %s count",
+                        rule[self.CLUSTER_NAME],
+                        rule[self.SERVICE_NAME],
+                        rule[self.ACTIVE_SIZE] if total_messages > 0 else 0,
+                    )
+                else:
+                    logger.info(
+                        "Service: %s %s is in state: %s; Desired count will not be adjusted",
+                        rule[self.CLUSTER_NAME],
+                        rule[self.SERVICE_NAME],
+                        current_state,
+                    )
 
                 self.log_queue_depth(queue_attributes, rule)
 
             except ClientError as ce:
                 logger.exception("Failed to update resources for rule %s", rule)
                 self.notify(ce, context=rule)
 
+    def get_description(self, rule):
+        """
+        Get the description of the service and its deployments - can be large
+        An unknown resource will result in an empty response.
+
+        The api specifies the ability to get the tags, but by observation they are not returned even when requested.
+        Use get_tags.
+        """
+        result = self.ecs.get_description(
+            cluster=rule[self.CLUSTER_NAME],
+            services=[rule[self.SERVICE_NAME]],
+            include=[
+                "TAGS"
+            ],  # appears to ignore this field - response does not have a tags field
+        )
```

Comment:
i wonder if https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ecs.html#ECS.Client.describe_services would work? not sure what get_description is tbh

Reply:
Yikes - good catch - that is a bad refactor - I am using describe_services. Unfortunately it does not appear to return tags even when requested.